### PR TITLE
feat/CAMPTASKS-395 : checking fetch calls for a const auth header

### PR DESCRIPTION
**Pull request by gersbach**
I'll squash this down once it is approved. Also {['Authorization']: blah} is another case that we should handle, but it might best to put that into a separate pull request.

### CODE REVIEW

**Comment by gersbach** [comment]
This currently doesn't handle the case insensitive case. Will fix and update.

**Comment by gersbach** [comment]
> This currently doesn't handle the case insensitive case. Will fix and update.

fixed!

File: crates/forge_analyzer/src/checkers.rs [hunk 1]
```diff
@@ -926,61 +927,60 @@ impl<'cx> Runner<'cx> for SecretChecker {
...
+                    ..
+                })) = operands.unwrap_or_default().get(1)
+                {
+                    let varid_argument =
+                        if let Some(Value::Object(varid)) = interp.get_value(def, *varid, None) {
+                            varid
+                        } else {
+                            varid
+                        };
+                    let headers_proj = projvec_from_str("Headers");
```
Comment (jwong101) [inline crates/forge_analyzer/src/checkers.rs:958]: The lowercase/uppercase normalization only applies to the actual headers themselves, not the headers property.

File: crates/forge_analyzer/src/checkers.rs [hunk 2]
```diff
@@ -926,61 +927,60 @@ impl<'cx> Runner<'cx> for SecretChecker {
...
+                    let headers_proj_lower = projvec_from_str("headers");
+                    if let Some(Value::Object(varid)) = interp
+                        .get_value(def, *varid_argument, Some(headers_proj))
+                        .or_else(|| {
+                            interp.get_value(def, *varid_argument, Some(headers_proj_lower))
+                        })
+                    {
+                        let auth_proj = projvec_from_str("Authorization");
+                        let aut_proj_lower = projvec_from_str("authorization");
+                        if let Some(Value::Const(_) | Value::Phi(_)) = interp
```
Comment (jwong101) [inline crates/forge_analyzer/src/checkers.rs:968]: `Value::Phi` should only be classified as a vuln if each of the values it can alias are also constants. Otherwise, the following would be classified as vuln, which shouldnt:

```javascript
  let tmp = process.env.PROD_SECRET;

  // also test for if the `authorization` property contains some uppercase or
  // lowercase letters (they all get normalized to lowercase, so they all map to the same header)
  let secret = "test";
  if (process.env.PROD) {
    secret = tmp;
  }

  fetch("url", { headers: { authorization: secret } });

```

File: crates/forge_analyzer/src/checkers.rs [hunk 3]
```diff
@@ -926,61 +927,60 @@ impl<'cx> Runner<'cx> for SecretChecker {
...
+                        };
+                    let headers_proj = projvec_from_str("Headers");
+                    let headers_proj_lower = projvec_from_str("headers");
+                    if let Some(Value::Object(varid)) = interp
+                        .get_value(def, *varid_argument, Some(headers_proj))
+                        .or_else(|| {
+                            interp.get_value(def, *varid_argument, Some(headers_proj_lower))
+                        })
+                    {
+                        let auth_proj = projvec_from_str("Authorization");
```
Comment (jwong101) [inline crates/forge_analyzer/src/checkers.rs:966]: Technically, any uppercase/lowercase combination of "authorization" works, however, these two are likely to be the only ones people use, so handling all cases can be a separate PR/task unless you want to do it in this one.

**Review by jwong101 — CHANGES_REQUESTED** [review]
Looks good! I have two concerns, but I like the overall approach.

File: crates/forge_analyzer/src/checkers.rs (hunk 1 above)
Comment (gersbach) [reply crates/forge_analyzer/src/checkers.rs]: fixed. Thank you!

File: crates/forge_analyzer/src/checkers.rs (hunk 2 above)
Comment (gersbach) [reply crates/forge_analyzer/src/checkers.rs:963]: This has been fixed. Phi is supposed to be a vec of constants, but this will no longer be a false positive.

File: crates/forge_analyzer/src/checkers.rs (hunk 3 above)
Comment (gersbach) [reply crates/forge_analyzer/src/checkers.rs:961]: should we loop through all of the characters, or do you think there is an alternative way to do this? Maybe make any vecs with properties authorization all lower case?

File: crates/forge_analyzer/src/interp.rs [hunk 4]
```diff
@@ -676,9 +564,154 @@ impl<'cx, C: Runner<'cx>> Interp<'cx, C> {
...
+        } = lval
+        {
+            let (varid, projections) = self.get_farthest_obj(defid_block, varid, projections);
+            let rval_value = self.value_from_rval(defid_block, rvalue);
+            if let Some(existing_lval) = self
+                .get_value(defid_block, varid, Some(projections.clone()))
+                .cloned()
+            {
+                // if there is an existing value...
+                match (existing_lval, rval_value) {
```
Comment (jwong101) [inline crates/forge_analyzer/src/interp.rs:587]: Don't forget to delete all subprojections when an object gets overwritten, otherwise the following test case will trigger:

```javascript
let obj = { headers: { authorization: 'secret' } };
obj = {};
fetch('test.com', obj);
```

**Review by jwong101 — CHANGES_REQUESTED** [review]

File: crates/forge_analyzer/src/interp.rs (hunk 4 above)
Comment (gersbach) [reply crates/forge_analyzer/src/interp.rs:587]: fixed. Thank you for the help! BTreeMap worked amazing!

File: crates/forge_analyzer/src/interp.rs [hunk 5]
```diff
@@ -676,9 +564,190 @@ impl<'cx, C: Runner<'cx>> Interp<'cx, C> {
...
+                            projections,
+                        ),
+                    (Value::Object(exist_var), Value::Object(new_var)) => {
+                        // store projection values that are transferred
+                        let mut projections_transferred = vec![];
+                        // tranfer all projection values from the new_var into the existing var
+                        for ((_, _, projections), value) in
+                            self.value_manager.varid_to_value.clone().range(
+                                (defid_block, new_var, ProjectionVec::new())
+                                    ..(defid_block, VarId(new_var.0 + 1), ProjectionVec::new()),
```
Comment (jwong101) [inline crates/forge_analyzer/src/interp.rs:627]: There's technically a bug in the event that `new_var.0 = u32::MAX`, since this would overflow back to 0. You can either create separate ranges for this case:

```rust
let start = (defid_block, new_var, smallvec![]);
let query = match new_var.0.checked_add(1) {
    Some(end) => start..(defid_block, VarId(end), smallvec![]),
    None => start..
};
for placeholder in self.value_manager.var_id_to_value.range(query) {
// etc
```

See [u32::checked_add](https://doc.rust-lang.org/std/primitive.u32.html#method.checked_add) for more details. Also, instead of cloning the entire `value_manager.varid_to_value`, you can instead create a `Vec` of the cloned `(value, projections)` using [Iterator::collect](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect) or `collect_vec`, and loop over them and call `self.add_value_with_projection` afterwards.

**Review by jwong101 — CHANGES_REQUESTED** [review]

File: crates/forge_analyzer/src/interp.rs [hunk 6]
```diff
@@ -676,9 +564,190 @@ impl<'cx, C: Runner<'cx>> Interp<'cx, C> {
...
+                            projections_transferred.push(projections.clone());
+                            self.add_value_with_projection(
+                                defid_block,
+                                exist_var,
+                                value.clone(),
+                                projections.clone(),
+                            )
+                        }
+                        // clear any remaining values
+                        for ((_, _, _), value) in self
```
[the remark this answers is not in the document]
Comment (gersbach) [reply crates/forge_analyzer/src/interp.rs:639]: Hmmm... I've spent some time trying to do this? Wouldn't we need to clone anyway since we are doing a query 2 times on a btreemap?

File: crates/forge_analyzer/src/interp.rs (hunk 5 above)
Comment (gersbach) [reply crates/forge_analyzer/src/interp.rs:627]: I've tried googling to get the end if the `checked_add(1)` is none. Is there anyway to do this? I couldn't see anyway to do this....

File: crates/forge_analyzer/src/interp.rs [hunk 7]
```diff
@@ -676,9 +564,190 @@ impl<'cx, C: Runner<'cx>> Interp<'cx, C> {
...
+                    .zip(all_values.iter())
+                    .flat_map(|(a, b)| vec![a.clone(), b.clone()])
+                    .chain(quasis_as_values.iter().skip(all_values.len()).cloned())
+                    .chain(all_values.iter().skip(quasis_as_values.len()).cloned())
+                    .collect_vec();
+
+                return_combinations_phi(values_joined)
+            }
+            Rvalue::Bin(BinOp::Add, op1, op2) => {
+                let value_op1 = self.value_from_operand(defid_block, op1);
```
[the remark this answers is not in the document]
Comment (gersbach) [reply crates/forge_analyzer/src/interp.rs:692]: fixed. Thank you! I am just going to exclude handling number tokens since none of our current scanners would require numbers....

File: crates/forge_analyzer/src/interp.rs [hunk 8]
```diff
@@ -676,9 +564,190 @@ impl<'cx, C: Runner<'cx>> Interp<'cx, C> {
...
+
+    #[inline]
+    fn get_farthest_obj(
+        &self,
+        defid_block: DefId,
+        varid: VarId,
+        mut projections: ProjectionVec,
+    ) -> (VarId, ProjectionVec) {
+        let mut current_var_id = varid;
+        for (i, _) in projections.clone().iter().enumerate() {
```
[the remark this answers is not in the document]
Comment (gersbach) [reply crates/forge_analyzer/src/interp.rs:738]: I needed to clone the projections here, not sure if there is a way around it, since we return the projections element.

File: crates/forge_analyzer/src/interp.rs (hunk 5 above)
Comment (jwong101) [reply crates/forge_analyzer/src/interp.rs:627]: You use a [one-sided range](https://doc.rust-lang.org/std/ops/struct.RangeFrom.html), since if `checked_add(1)` is `None`, then you just need the rest of the map. The one-sided range is `start..` in this case.

File: crates/forge_analyzer/src/interp.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (jwong101) [reply crates/forge_analyzer/src/interp.rs:738]: Not for the loop, since you're only using the index. Instead of doing `for (i, _) in projections.iter().enumerate()`, you can just loop over the indices directly, i.e. `0..projections.len()`.

File: crates/forge_analyzer/src/interp.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jwong101) [reply crates/forge_analyzer/src/interp.rs:639]: Yeah, but you only need to clone the `new_var`'s `Value`s and `Projection`s. For instance:

```rust
    for ((_, _, _), value) in self.value_manager.varid_to_value.range_mut(
        (defid_block, exist_var, ProjectionVec::new())
            ..(defid_block, VarId(exist_var.0 + 1), ProjectionVec::new()),
    ) {
        *value = Value::Unknown
    }
    // tranfer all projection values from the new_var into the existing var
    let vals = self
        .value_manager
        .varid_to_value
        .range(
            (defid_block, new_var, ProjectionVec::new())
                ..(defid_block, VarId(new_var.0 + 1), ProjectionVec::new()),
        )
        .map(|((_, _, projections), value)| (projections.clone(), value.clone()))
        .collect_vec();
    for (projections, value) in vals {
        self.add_value_with_projection(defid_block, exist_var, value, projections)
    }
```

File: crates/forge_analyzer/src/interp.rs (hunk 5 above)
Comment (gersbach) [reply crates/forge_analyzer/src/interp.rs:627]: I am getting this compilation error. Do I need to find the max element of the var and the max element of the projection vec? Sorry for so many questions.... 

I was trying to find a way to solve it, but I was unsuccessful 😅 

```
25 |                           let query = match new_var.0.checked_add(1) {
    |  _____________________________________-
626 | |                             Some(end) => start..(defid_block, VarId(end), ProjectionVec::new()),
    | |                                          ------------------------------------------------------ this is found to be of type `std::ops::Range<(DefId, VarId, SmallVec<[Projection; 1]>)>`
627 | |                             None => start..
    | |                                     ^^^^^^^ expected `Range<(DefId, VarId, ...)>`, found `RangeFrom<(DefId, VarId, ...)>`
628 | |                         };
    | |_________________________- `match` arms have incompatible types
    |
```

File: crates/forge_analyzer/src/interp.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (gersbach) [reply crates/forge_analyzer/src/interp.rs:738]: fixed. Thank you!

File: crates/forge_analyzer/src/interp.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (gersbach) [reply crates/forge_analyzer/src/interp.rs:639]: When we add the values first, we call `self.get_farthest_obj()`. However, when we use the BTreeMap first, we don't check this instead assigning all of them to unknown. As a result, we cannot access the object via `self.get_farthest_object` since it is overwritten with unknown. But, I will implement the cloning only for the projections in the first one. Thank you!

File: crates/forge_analyzer/src/interp.rs (hunk 5 above)
Comment (jwong101) [reply crates/forge_analyzer/src/interp.rs:627]: Ah, my bad. You need to put the queries in the different arms like so: 

```rust
let start = (defid_block, new_var, smallvec![]);
let query = match new_var.0.checked_add(1) {
    Some(end) => self.value_manager.var_id_to_value.range(start..(defid_block, VarId(end), smallvec![])),
    None => self.value_manager.var_id_to_value.range(start..),
};
for placeholder in query {
// etc
```

I forgot that `x..y` and `x..` are two different types.

File: crates/forge_analyzer/src/interp.rs [hunk 9]
```diff
@@ -676,9 +564,203 @@ impl<'cx, C: Runner<'cx>> Interp<'cx, C> {
...
         projections: ProjectionVec,
     ) {
+        let (varid, projections) = self.get_farthest_obj(defid_block, varid, projections);
         self.value_manager
             .insert_var_with_projection(defid_block, varid, projections, value);
     }
+
+    // this function takes in an operand checks for previous values and returns a value optional
+    #[inline]
+    pub fn add_value_to_definition(&mut self, defid_block: DefId, lval: Variable, rvalue: Rvalue) {
```
Comment (jwong101) [inline crates/forge_analyzer/src/interp.rs:574]: Looks good! There are two edge cases that you need to handle though:

There's a bug where we set the underlying object's properties to `Unknown` if it's assigned to itself.

There's another bug where the alias isn't being updated if the property was an object. I believe this is because you're only getting the underlying object via `get_farthest_obj`, but not updating properties that may exist in the alias.

```javascript
let h = { headers: { authorization: 'foo' } };
h = h;
// Bug 1: Should trigger a vuln, currently doesn't
fetch('url', h)

// Bug 2: Shouldn't trigger a vuln, currently does
let c = h;
c.headers = {};
fetch('url', h);
```

Note that you have to be careful, since aliases are only updated in Javascript if a property is used, i.e:

```javascript
let c = h.headers;
// h.headers is still { authorization: 'foo' } after this assignment.
c = {};
```

**Review by jwong101 — CHANGES_REQUESTED** [review]

File: crates/forge_analyzer/src/interp.rs (hunk 9 above)
Comment (gersbach) [reply crates/forge_analyzer/src/interp.rs:574]: fixed! Thank you!
